### ISSUE TITLE
Ignores variable type hint for union of tuples

### ISSUE THREAD

**Issue by trygveaa** [microsoft/pyright#4998]
**Describe the bug**
If I have a value with type `tuple[float, ...]` and assign it to a variable with a type hint for a union of tuples with a specific length, pyright ignores the variable type hint without any error.

**To Reproduce**
Run pyright on this code:

```python
x: tuple[float, float] | tuple[float, float, float] = tuple([1.0, 2.0])
reveal_type(x)
```

Reveals that the type of `x` is `tuple[float]`.

**Expected behavior**
 I expected pyright to honor my type hint so the type of `x` would be `tuple[float, float] | tuple[float, float, float]`, or alternatively report an error if the type isn't compatible.

**VS Code extension or command-line**
Via command-line, version 1.1.304.

**Comment by erictraut** [comment]
There is a bug here, but the correct behavior is not quite what you expect.

The evaluated type of `[1.0, 2.0]` is `list[float]`, and the evaluated type of `tuple([1.0, 2.0])` is `tuple[float, ...]`. This type is compatible with both `tuple[float, float]` and `tuple[float, float, float]` because it is an unknown length. The imprecise `...` form of `tuple` was added for incremental type checking. It is similar to `Callable[..., None]`, which is compatible with a callable type with any signature.

The current type evaluation logic is producing a result of `tuple[float]`, which is incorrect. It should be `tuple[float, ...]`. There should be no type error reported in this case.

I recommend avoiding the use of the `tuple` constructor and using a tuple expression where possible because the constructor call often obscures the length of the tuple, which can lead to undetected errors like in this case.

This fix will be included in the next release.

**Comment by trygveaa** [comment]
Thanks. It makes sense that `float` is narrowed to `int` (though, it can be a bit surprising that the type isn't what I specified it to be). However, why is `tuple[float, float] | tuple[float, float, float]` "narrowed" to `tuple[float, ...]` when the latter is a wider type?

When I declare `x` with the type `tuple[float, float] | tuple[float, float, float]` I don't expect pyright to allow me to call `x[2]`. However, when I assign the result of `tuple(...)`, `x` is changed/narrowed to `tuple[float, ...]`, and then suddenly `x[2]` is allowed.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.305, which I just published. It will also be included in a future release of pylance.
